autoload/maktaba/syscall.vim

```diff
@@ -89,6 +109,58 @@ function! maktaba#syscall#DoCall() abort dict
   return l:return_data
 endfunction
 
+
+""
+" @private
+" @dict Syscall
+" Calls |system()| asynchronously, and invokes a @function(this.callback) once
+" the command completes, passing in stdout, stderr and exit code to it.
+" The specific implementation for @function(#CallAsync).
+function! maktaba#syscall#DoCallAsync() abort dict
+  if empty(v:servername)
+    throw maktaba#error#Message('ShellError', 'Cannot run async commands, no' .
+        \ ' --servername flag passed to vim. See :help servername.')
+  elseif !has('clientserver')
+    throw maktaba#error#Message('ShellError', 'Cannot run async commands, ' .
+        \ 'vim was compiled without +clientserver. See :help clientserver')
+  endif
+
+  let l:error_file = tempname()
+  let l:output_file = tempname()
+  let l:callback_cmd = join([
+        \ v:progname,
+        \ '--servername ' . v:servername,
+        \ '--remote-expr',
+        \ '"maktaba#syscall#AsyncDone(',
+        \ "'" . l:output_file . "', ",
+        \ "'" . l:error_file . "', ",
+        \ '$?',
+        \ ')"'], " ")
+  let l:full_cmd = printf('(%s; %s >/dev/null) > %s 2> %s &',
+        \ self.GetCommand(), l:callback_cmd, l:output_file, l:error_file)
+  let s:callbacks[l:output_file] = {
+        \ 'function': maktaba#ensure#IsCallable(self.callback),
+        \ 'tab': tabpagenr()}
+
+  let l:shell_state = maktaba#value#SaveAll(['&shell', '$SHELL'])
+  if &shell !~# s:usable_shell
+    set shell=/bin/sh
+  endif
+  if $SHELL !~# s:usable_shell
+    let $SHELL = '/bin/sh'
+  endif
+
+  try
+    silent execute '! ' . s:EscapeSpecialChars(l:full_cmd)
```

Comment:
Any reason to use `:!` instead of `system()`? I'd lean towards `system()` in scripts unless there's a specific reason to execute more interactively.